tests/unit/test_ttl_cache.py

```diff
@@ -0,0 +1,292 @@
+import random
+from concurrent.futures import ThreadPoolExecutor, wait
+from unittest.mock import patch
+
+import maya
+import pytest
+
+from nucypher.utilities.cache import TTLCache
+
+
+def test_cache_invalid_ttl():
+    with pytest.raises(ValueError):
+        TTLCache(ttl=0)
+
+    with pytest.raises(ValueError):
+        TTLCache(ttl=-1)
+
+
+def test_cache_invalid_key_value_pair():
+    ttl_cache = TTLCache(ttl=60)
+    with pytest.raises(ValueError):
+        ttl_cache[None] = "a"
+
+    with pytest.raises(ValueError):
+        ttl_cache["a"] = None
+
+
+def test_cache_get_remove_non_existent_entry():
+    ttl_cache = TTLCache(ttl=60)
+
+    # try to get non-existent entry
+    assert ttl_cache[1] is None
+
+    # try to remove non-existent entry
+    ttl_cache.remove(1)
```

Comment:
Sorry, I don't get how this line works. Should be captured here an exception or something like that about we are trying to delete a non-existent entry?

Reply:
Fair question. 

It's really ensuring that a `KeyError` isn't raised.

Remember that cache entries can be removed (in the backgrounds) if they are expired, so for `remove()`, one of the following scenarios is possible:
a) Ask to remove a keyed entry that is currently in the cache
b) Ask to remove a keyed entry that was never in the cache
c) Ask to remove a keyed entry that is in the cache but expired
d) Ask to remove a keyed entry that was in the cache but was auto-removed previously because it was expired (eg. someone called length on the cache or something that performed a purge)

- a) is the regular case - so remove the item, great
- for b) we could throw a `KeyError` but how do we tell the difference between b) and scenario d)
- for c) would you do nothing or `KeyError`? If `KeyError` what's the difference between scenario c) and d) - just the timing of when you happen to call `remove()`? That's pretty inconsistent.

So to be easily consistent, we just never return a `KeyError` and instead do nothing if someone tries to remove a non-existent entry. That philosophy holds for the cache in general i.e. no `KeyError` is raised - same for the `__getitem__()` functionality.

Reply:
Got it, thanks for your nice clarification